### ISSUE TITLE
Graphql price tiers returning data from previous item in foreach

### ISSUE THREAD

**Issue by treestonemedia** [magento/magento2#32279]
### Preconditions 

1. Clean Magento 2.4.2 with sample data
1. 2.4-develop

### Steps to reproduce (*)

1. Install clean Magento 2.4.2 with sample data
1. Add tier pricing to any item (for this example we added tier prices to items ID 1 & 2)
1. run Graphql query with search that will include the item with tier prices and items without tier prices
```
{
  products(pageSize: 20, currentPage: 1, search: "24-MB0") {
    items {
      id
      price_tiers {
        quantity
        final_price {
          value
        }
      }
    }
  }
}
{code}

### Expected result (*)

1. Items with tier price return correct tier pricing, items without tier pricing return no tier prices
{code}
{{
{
    "data": {
        "products": {
            "items": <
                {
                    "id": 42,
                    "price_tiers": []
                },
				...
                {
                    "id": 17,
                    "price_tiers": []
                },
                {
                    "id": 1,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 23,
                    "price_tiers": []
                },
                {
                    "id": 19,
                    "price_tiers": []
                },
				...
            ]
        }
    }
}
}}
{code}

### Actual result (*)

1. All items coming after an item with tier pricing have tier pricing in the response
1. The tier pricing data that it's returning is from the last item in the loop that had tier pricing
1. in our example, items ID 1 & ID 2 are the only ones that have tier pricing, yet here is the result

{code}
{
    "data": {
        "products": {
            "items": [
                {
                    "id": 42,
                    "price_tiers": []
                },
                {
                    "id": 2046,
                    "price_tiers": []
                },
                {
                    "id": 11,
                    "price_tiers": []
                },
                {
                    "id": 12,
                    "price_tiers": []
                },
                {
                    "id": 6,
                    "price_tiers": []
                },
                {
                    "id": 3,
                    "price_tiers": []
                },
                {
                    "id": 41,
                    "price_tiers": []
                },
                {
                    "id": 18,
                    "price_tiers": []
                },
                {
                    "id": 39,
                    "price_tiers": []
                },
                {
                    "id": 17,
                    "price_tiers": []
                },
                {
                    "id": 1,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 23,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 19,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 8,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 13,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 15,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 16,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 22,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 43,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                },
                {
                    "id": 44,
                    "price_tiers": [
                        {
                            "quantity": 1,
                            "final_price": {
                                "value": 34
                            }
                        },
                        {
                            "quantity": 5,
                            "final_price": {
                                "value": 32
                            }
                        }
                    ]
                }
            >
        }
    }
}
```

1. The issue started happening since commit https://github.com/magento/magento2/commit/b2021d55c022719af7adec3f5e5fb6b1776e6fef#diff-8d1c1372668a97ccd40b222f73acfe4c6b5ce24ad261215306610ae8817cdc65 

cc: @Usik2203 

## 
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- <x] Severity: **S0** *- Affects critical data or functionality and leaves users without workaround.*
- [ ] Severity: **S1** *- Affects critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S2** *- Affects non-critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S3** *- Affects non-critical data or functionality and does not force users to employ a workaround.*
- [ > Severity: **S4** *- Affects aesthetics, professional look and feel, “quality” or “usability”.*

**Comment by treestonemedia** [comment]
@Usik2203 can you please take a look at this?

**Comment by engcom-Bravo** [comment]
Hello @treestonemedia 

Thank you for your report.

We have reproduced the reported issue on the latest 2.4-develop. 
After adding tier price to the products _24-MB01_ and _24-MB04_, the GraphQL search query returns tier prices for products that have not been changed.
**Before** adding tier price to products _24-MB01_ and _24-MB04_
```graphql
{
  "data": {
    "products": {
      "items": [
        {
          "id": 43,
          "price_tiers": []
        },
        {
          "id": 2047,
          "price_tiers": []
        },
        {
          "id": 12,
          "price_tiers": []
        },
        {
          "id": 13,
          "price_tiers": []
        },
        {
          "id": 7,
          "price_tiers": []
        },
        {
          "id": 4,
          "price_tiers": []
        },
        {
          "id": 42,
          "price_tiers": []
        },
        {
          "id": 2,
          "price_tiers": []
        },
        {
          "id": 19,
          "price_tiers": []
        },
        {
          "id": 40,
          "price_tiers": []
        },
        {
          "id": 18,
          "price_tiers": []
        },
        {
          "id": 20,
          "price_tiers": []
        },
        {
          "id": 24,
          "price_tiers": []
        },
        {
          "id": 23,
          "price_tiers": []
        },
        {
          "id": 44,
          "price_tiers": []
        },
        {
          "id": 17,
          "price_tiers": []
        },
        {
          "id": 14,
          "price_tiers": []
        },
        {
          "id": 9,
          "price_tiers": []
        },
        {
          "id": 16,
          "price_tiers": []
        },
        {
          "id": 37,
          "price_tiers": []
        }
      ]
    }
  }
}
```
**After** adding tier price to products _24-MB01_ and _24-MB04_
```graphql
{
  "data": {
    "products": {
      "items": [
        {
          "id": 43,
          "price_tiers": []
        },
        {
          "id": 2047,
          "price_tiers": []
        },
        {
          "id": 12,
          "price_tiers": []
        },
        {
          "id": 13,
          "price_tiers": []
        },
        {
          "id": 7,
          "price_tiers": []
        },
        {
          "id": 4,
          "price_tiers": []
        },
        {
          "id": 42,
          "price_tiers": []
        },
        {
          "id": 19,
          "price_tiers": []
        },
        {
          "id": 40,
          "price_tiers": []
        },
        {
          "id": 2,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 18,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 20,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 24,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 23,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 44,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 17,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 14,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 9,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 16,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        },
        {
          "id": 37,
          "price_tiers": [
            {
              "quantity": 2,
              "final_price": {
                "value": 10
              }
            }
          ]
        }
      ]
    }
  }
}
```

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-41084` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by Usik2203** [comment]
@magento I am working on this

**Comment by SirCoolness** [comment]
@magento give me 2.4.2 instance

**Comment by gabrieldagama** [comment]
Hi @treestonemedia. Thank you for your report. 
The issue has been fixed in magento/magento2#32353 by @Usik2203 in 2.4-develop branch
Related commit(s):
 - [3359cfa671573656d9a6754397160609389af365](https://github.com/magento/magento2/commit/3359cfa671573656d9a6754397160609389af365)
 - [706d2ec8829a8ab82d426a4b9447212ea80f293a](https://github.com/magento/magento2/commit/706d2ec8829a8ab82d426a4b9447212ea80f293a)
 - [457b0ea97f7c05e8d8fe20ab8de4ef924cddff6d](https://github.com/magento/magento2/commit/457b0ea97f7c05e8d8fe20ab8de4ef924cddff6d)
 - [d5c419115ef738ed6c02403dcf8184032df3e308](https://github.com/magento/magento2/commit/d5c419115ef738ed6c02403dcf8184032df3e308)
 - [022d7bf598ea584a5423b3dc58372463df0a7579](https://github.com/magento/magento2/commit/022d7bf598ea584a5423b3dc58372463df0a7579)
 - [60e258ff013ebb3f71c93cc44b29c87264679e44](https://github.com/magento/magento2/commit/60e258ff013ebb3f71c93cc44b29c87264679e44)
 - [f494cb1b43cb520143b2c54091b7a9874ad1797c](https://github.com/magento/magento2/commit/f494cb1b43cb520143b2c54091b7a9874ad1797c)
 - [5590d7c3757ff425519272c2df39e6f05d03b8de](https://github.com/magento/magento2/commit/5590d7c3757ff425519272c2df39e6f05d03b8de)

The fix will be available with the upcoming 2.4.3 release.
